File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private List<Integer> path = new LinkedList<>();
    private List<List<Integer>> ans = new ArrayList<>();
    
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtracking(candidates, 0, target, 0);
        return ans;
    }
    
    private void backtracking(int[] candidates, int sum, int targetSum, int startIdx) {
        if (sum == targetSum) {
            ans.add(new ArrayList<>(path));
            return;
        }
        
        for (int i = startIdx; i < candidates.length; i++) {
            if (i > startIdx && candidates[i] == candidates[i - 1]) {
                continue;
            }
            
            if (sum > targetSum) {
                break;
            }
            
            sum += candidates[i];
            path.add(candidates[i]);
            backtracking(candidates, sum, targetSum, i + 1);
            path.removeLast();
            sum -= candidates[i];
        }
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        # 1. sort the candidates array to avoid duplicate combination
        candidates.sort()
        combinations = []

        def backtracking(index: int, path: List[int], curr_sum) -> None:
            if curr_sum == target:
                combinations.append(path[:])
                return
            
            for i in range(index, len(candidates)):
                if curr_sum > target:
                    break
                
                if i > index and candidates[i] == candidates[i - 1]:
                    continue

                path.append(candidates[i])
                curr_sum += candidates[i]
                backtracking(i + 1, path, curr_sum)
                curr_sum -= candidates[i]
                path.pop()
        
        backtracking(0, [], 0)
        return combinations